plugins/flytekit-async-fsspec/flytekitplugins/async_fsspec/s3fs/s3fs.py

```diff
@@ -0,0 +1,245 @@
+
+                # async function to upload a single chunk
+                async def upload_chunk(chunk, part_number):
+                    result = await self._call_s3(
+                        "upload_part",
+                        Bucket=bucket,
+                        PartNumber=part_number,
+                        UploadId=mpu["UploadId"],
+                        Body=chunk,
+                        Key=key,
+                    )
+                    callback.relative_update(len(chunk))
+                    return {"PartNumber": part_number, "ETag": result["ETag"]}
+
+                tasks = set()
+                part_number = 1
+                parts = []
+                read_end = False
+                while True:
+                    while len(tasks) < concurrent_upload:
+                        chunk = f0.read(chunksize)
+                        if not chunk:
+                            read_end = True
+                            break
+                        tasks.add(asyncio.create_task(upload_chunk(chunk, part_number)))
+                        part_number += 1
+                    if read_end:
+                        break
+                    done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+                    parts.extend(map(lambda x: x.result(), done))
+                    tasks = pending
+
+                parts.extend(await asyncio.gather(*tasks))
+                parts.sort(key=lambda part: part["PartNumber"])
+                await self._call_s3(
+                    "complete_multipart_upload",
+                    Bucket=bucket,
+                    Key=key,
+                    UploadId=mpu["UploadId"],
+                    MultipartUpload={"Parts": parts},
+                )
+        while rpath:
+            self.invalidate_cache(rpath)
+            rpath = self._parent(rpath)
+
+    async def _get_file(
+        self,
+        rpath,
+        lpath,
+        callback=_DEFAULT_CALLBACK,
+        version_id=None,
+        chunksize=DEFAULT_DOWNLOAD_CHUNK_SIZE,
+        concurrent_download=DEFAULT_CONCURRENT_DOWNLOAD,
+    ):
+        """
+        Get a file from rpath to lpath.
+        Args:
+            rpath (str): The remote path of the file to be downloaded.
+            lpath (str): The local path which the file should be downloaded to.
+            callback (function, optional): The callback function.
+            chunksize (int, optional): Download chunksize. Defaults to 50 * 2**20 (50MB).
+            version_id (str, optional): The version id of the file. Defaults to None.
+        """
+        if os.path.isdir(lpath):
+            return
+
+        # get the file size
+        file_info = await self._info(path=rpath, version_id=version_id)
+        file_size = file_info["size"]
+
+        bucket, key, vers = self.split_path(rpath)
+
+        # the async function to get a range of the remote file
+        async def _open_file(start_byte: int, end_byte: int = None):
```

Comment:
can we change it so that we handle the errors here and retries here. the lower logic will then be simplified

Reply:
I have made the error handling a function and the lower logic has been simplified. Thank you for the suggestion.

Reply:
I made the error handling a function and the lower logic got cleaner. Thank you for the suggestion.